csharp/src/Google.Protobuf/Reflection/DescriptorPool.cs

```diff
@@ -189,13 +186,38 @@ private static void ValidateSymbolName(IDescriptor descriptor)
             {
                 throw new DescriptorValidationException(descriptor, "Missing name.");
             }
-            if (!ValidationRegex.IsMatch(descriptor.Name))
+
+            if (!IsValidSymbolNameFormat(descriptor.Name))
             {
                 throw new DescriptorValidationException(descriptor,
                                                         "\"" + descriptor.Name + "\" is not a valid identifier.");
             }
         }
 
+        /// <summary>
+        /// Symbol name must start with a letter or underscore, and it can contain letters, numbers and underscores.
+        /// </summary>
+        private static bool IsValidSymbolNameFormat(string name)
+        {
+            if (!IsAsciiLetter(name[0]) && name[0] != '_')
+            {
+                return false;
+            }
+
+            for (int i = 1; i < name.Length; i++)
+            {
+                if (!IsAsciiLetter(name[i]) && !IsAsciiDigit(name[i]) && name[i] != '_')
+                {
+                    return false;
+                }
+            }
+
+            return true;
+
+            static bool IsAsciiLetter(char c) => (uint) ((c | 0x20) - 'a') <= 'z' - 'a';
+            static bool IsAsciiDigit(char c) => (uint) (c - '0') <= ('9' - '0');
```

Comment:
```suggestion
            static bool IsAsciiDigit(char c) => (uint) (c - '0') <= '9' - '0';
```

(Or add the parentheses to IsAsciiLetter.)